Allow ref override

alanmcruickshank opened the pull request:
This is a very late re-entry, but the bug has hit me again and I thought I should get round to resubmitting a fox for #1603 .

These edits allow users to override built in functions like `ref`.

For our use case (where we want to stop rendering the `database` portion of the object reference), the user just needs to have a macro such as:

`ref_override.sql`:

```jinja
{% macro ref(modelname) %}{{ builtins.ref(modelname).include(database=False).render() }}{% endmacro %}
```

And the ref function will be overridden, now not rendering the `database`.

@drewbanin - I think this is much closer to what you originally envisioned. Have I overridden the parts that you had originally intended? If this is in line, I'd like to add the example macro above somewhere in the documentation for other users to find. Where would the best place be?

alanmcruickshank commented:
@drewbanin - I've now signed the CLA.

drewbanin commented:
Thanks for making this PR @alanmcruickshank! I think this looks really good - I'll do some functional testing now. Adding @beckjake to review since he's spent a lot of time around this part of the codebase recently.

This probably won't make it out for 0.15.1 - can you change the base branch to `dev/barbara-gittings`? Thanks also for signing the CLA :) @cla-bot check

beckjake commented on core/dbt/context/common.py:
```diff
@@ -211,16 +211,22 @@ def search_package_name(self):
+                self.db_wrapper, self.model, self.config, self.manifest),
+            config=self.provider.Config(
+                self.model, self.source_config),
+            execute=self.provider.execute
```
This is a nit, but can you please use the explicit dict construction syntax like:

```
{
    'ref': self.provider.ref(...),
    'source': ...,
    ...
}
```

instead?

alanmcruickshank replied:
Sure. I'm not particularly fussy between the two syntax constructions.

beckjake commented on core/dbt/context/common.py:
```diff
@@ -211,16 +211,22 @@ def search_package_name(self):
+        # Install each of them directly in case they're not
+        # clobbered by a macro.
+        for builtin in builtins:
+            context[builtin] = builtins[builtin]
```
This is equivalent to `context.update(builtins)`, right? I find that a bit clearer, IMO the only reason to use a loop here is if you need to update without replacement (and in that case, I'd iterate over `.items()`).

alanmcruickshank replied:
👍

beckjake commented on core/dbt/context/base.py:
```diff
@@ -248,9 +249,14 @@ def add_macros_from(
             _add_macro_map(context, package_name, macro_map)
 
-            if package_name == self.search_package_name:
+            if package_name == config.project_name:
```
Instead of changing the method signature, please just use `self.config.project_name` here!

alanmcruickshank replied:
Ah, I didn't realise we have a `config` object in the class already. Great idea.

alanmcruickshank commented:
Thanks @drewbanin @beckjake .

- Rebased onto `dev/barbara-gittings`.
- Added changes from code review. All the suggestions were very sensible. Thanks @beckjake.
- I'd love to write some integration tests around this too. Last time I tried I found it really hard to work out how to write tests for the rendering of templates. @beckjake is there an existing test that would be a good starting point to write some tests for this? With an existing example and a couple of pointers, I'm happy to do the rest of the legwork.

beckjake commented:
@alanmcruickshank My method for writing a new integration test is usually this:

1) make a new folder with `${test_folder_number}_${somewhat_useful_name}` under `test/integration`, where `test_folder_number=${biggest_existing_number} + 1` - in this case, probably something like `055_ref_override_test`. It doesn't have to be great, but I should be able to figure out what kind of tests are in there by reading it.
2) Make a new file in that folder named `test_${something}.py`. `test_ref_override.py` is probably fine.
4) Put something following this pattern in there:
```
from test.integration.base import DBTIntegrationTest, use_profile


class TestRefOverride(DBTIntegrationTest):
    @property
    def schema(self):
        return "${some_name}_${test_folder_number}"

    # you'll need this if you're using seeds/data tests in your test
    @property
    def project_config(self):
        return {
            'data-paths': ['data'],
            'test-paths': ['tests'],
        }

    @property
    def models(self):
        return "models"

    @use_profile('postgres')
    def test_postgres_ref_override(self):
        self.run_dbt(['seed']) #if this is how you're testing
        self.run_dbt(['run'])
        self.run_dbt(['test']) #if this is how you're testing
        self.assertTablesEqual('seed_model', 'my_model')
```

The test should have `_postgres_` in the method name and have the `@use_profile('postgres')` decorator. If you don't have both, the test suite will probably complain. By default, `run_dbt` asserts that `dbt ${cmd}` exited successfully.

5) Make a `models` directory in your test folder, and put any relevant test models in there that exercise your change, and potentially any tests that match it in a `schema.yml` file. You can also add `macros` and `data` directories alongside that contain macros (I assume you'll need that!) or seeds. Similarly, `tests` if you need data tests.

For the test itself, there are broadly two techniques:
 - have a seed that creates a table you want, and assert that the model `dbt seed` creates is equal to the model `dbt run` creates. Instead of using `dbt seed` you can write a seed sql file that you execute with `self.run_sql_file()` as part of the test, but tests using `dbt seed` are usually easier to maintain/migrate to new databases.
 - have a schema test/data test that asserts your run did the right thing.

`009_data_tests_test` is a pretty god example of the second technique, I guess `005_simple_seed_test` is a decent example of the first technique (see the `test_column_type` macro). A lot of tests kind of blend the two methods.

beckjake commented on test/integration/055_ref_override_test/test_ref_override.py:
```diff
@@ -0,0 +1,26 @@
+        return {
+            'data-paths': ['data'],
+            "macro-paths": ["macros"],
+        }
```
you probably need a `'quote_columns': False,` entry here too, to quiet that deprecation warning.

beckjake commented:
I should have mentioned: I haven't set it up fresh in quite a while, but if you don't want to push to trigger tests, and you have docker/docker-compose set up, you should be able to run your test locally with something like this:

```
cp test.env.sample test.env
docker-compose run --rm test tox -e explicit-py36 -- -s -x -m profile_postgres test/integration/055_ref_override_test
```


The first time will probably take a while as it has to pull some big silly containers (postgres + our test one) and build the tox environment.

beckjake commented on test/integration/055_ref_override_test/macros/ref_override_macro.sql:
```diff
@@ -0,0 +1,2 @@
+-- Macro to override ref and always return the same result
+{% macro ref(modelname) %} seed_2 {% endmacro %}
```
Here's a suggestion for a perhaps more interesting macro that exercises the code more completely:

```suggestion
{% macro ref(modelname) %}
{% do return(builtins.ref(modelname).replace_path(identifier='seed_2')) %}
{% endmacro %}

```

alanmcruickshank commented:
@beckjake - integration test is in and working - thanks for the pointers! Yes it did take a while to download but still faster than the azure tests 😄 .

I've squashed all my commits for tidiness. Ready for review I think.

drewbanin commented:
just kicked off the tests here @alanmcruickshank :)

drewbanin commented:
Really nice work @alanmcruickshank! I think we'll want to update the docs here: https://docs.getdbt.com/docs/jinja-context

We can add a section for `builtins` and show an example that uses this macro under the `ref` entry. My understanding is that you wanted this feature for a sort of blue/green deployment scheme, right? It would be awesome to write up a guide on how to do something like that, then link out to these docs for the specifics.

I usually handle the docs updates when a new release is ready, but let me know if there are specific things you'd like to touch on and I can be sure to include them.

Merging this! Thanks again for this really lovely PR :)
